File: samples/VIN/java/VINScanner/src/main/java/VINScanner.java
import com.dynamsoft.core.EnumErrorCode;
import com.dynamsoft.core.basic_structures.FileImageTag;
import com.dynamsoft.core.basic_structures.ImageTag;
import com.dynamsoft.cvr.CaptureVisionRouter;
import com.dynamsoft.cvr.CapturedResult;
import com.dynamsoft.dcp.EnumValidationStatus;
import com.dynamsoft.dcp.ParsedResult;
import com.dynamsoft.dcp.ParsedResultItem;
import com.dynamsoft.license.LicenseError;
import com.dynamsoft.license.LicenseException;
import com.dynamsoft.license.LicenseManager;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Objects;
import java.util.Scanner;

class VINResult {
    String codeType;
    String wmi;
    String region;
    String vds;
    String vis;
    String modelYear;
    String plantCode;
    String vinString;

    public VINResult(ParsedResultItem item) {
        codeType = item.getCodeType();
        if (!Objects.equals(codeType, "VIN")) {
            return;
        }
        if (item.getFieldValue("vinString") != null) {
            vinString = item.getFieldValue("vinString");
            if (item.getFieldValidationStatus("vinString") == EnumValidationStatus.VS_FAILED) {
                vinString += ", Validation Failed";
            }
        }            if (item.getFieldValue("WMI") != null && item.getFieldValidationStatus("WMI") != EnumValidationStatus.VS_FAILED) {
            wmi = item.getFieldValue("WMI");
        }
        if (item.getFieldValue("region") != null && item.getFieldValidationStatus("region") != EnumValidationStatus.VS_FAILED) {
            region = item.getFieldValue("region");
        }
        if (item.getFieldValue("VDS") != null && item.getFieldValidationStatus("VDS") != EnumValidationStatus.VS_FAILED) {
            vds = item.getFieldValue("VDS");
        }
        if (item.getFieldValue("modelYear") != null && item.getFieldValidationStatus("modelYear") != EnumValidationStatus.VS_FAILED) {
            modelYear = item.getFieldValue("modelYear");
        }
        if (item.getFieldValue("plantCode") != null && item.getFieldValidationStatus("plantCode") != EnumValidationStatus.VS_FAILED) {
            plantCode = item.getFieldValue("plantCode");
        }
    }

    @Override
    public String toString() {
        return "VIN String: " + vinString + "\n" +
                "Parsed Information:\n" +
                "\tWMI: " + (wmi != null ? wmi : "") + "\n" +
                "\tRegion: " + (region != null ? region : "") + "\n" +
                "\tVDS: " + (vds != null ? vds : "") + "\n" +
                "\tVIS: " + (vis != null ? vis : "") + "\n" +
                "\tModelYear: " + (modelYear != null ? modelYear : "") + "\n" +
                "\tPlantCode: " + (plantCode != null ? plantCode : "") + "\n";
    }
}

public class VINScanner {
    private static void printResults(ParsedResult result) {
        ImageTag tag = result.getOriginalImageTag();
        if (tag instanceof FileImageTag) {
            System.out.println("File: " + ((FileImageTag)tag).getFilePath());
        }

        if (result.getErrorCode() != EnumErrorCode.EC_OK && result.getErrorCode() != EnumErrorCode.EC_UNSUPPORTED_JSON_KEY_WARNING) {
            System.out.println("Error: " + result.getErrorString());
        } else {
            ParsedResultItem[] items = result.getItems();
            System.out.println("Parsed " + items.length + " VIN codes.");
            for (ParsedResultItem item : items) {
                VINResult vinResult = new VINResult(item);
                System.out.println(vinResult);
            }
        }
    }

    public static void main(String[] args) {
        System.out.println("**********************************************************");
        System.out.println("Welcome to Dynamsoft Capture Vision - VIN Sample");
        System.out.println("**********************************************************");

        Scanner scanner = new Scanner(System.in);

        try {
            int errorCode = 0;
            String errorMsg = "";

            // Initialize license.
            // You can request and extend a trial license from https://www.dynamsoft.com/customer/license/trialLicense?product=dcv&utm_source=samples&package=java
            // The string 'DLS2eyJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSJ9' here is a free public trial license. Note that network connection is required for this license to work.
            try {
                LicenseError licenseError = LicenseManager.initLicense("DLS2eyJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSJ9");
                if (licenseError.getErrorCode() != EnumErrorCode.EC_OK) {
                    errorCode = licenseError.getErrorCode();
                    errorMsg = licenseError.getErrorString();
                }
            } catch (LicenseException e) {
                errorCode = e.getErrorCode();
                errorMsg = e.getErrorString();
            }

            if (errorCode != EnumErrorCode.EC_OK) {
                System.out.println("License initialization failed: ErrorCode: " + errorCode + ", ErrorString: " + errorMsg);
                System.out.print("Press Enter to quit...");
                scanner.nextLine();
                return;
            }

            CaptureVisionRouter cvRouter = new CaptureVisionRouter();
            while (true) {
                System.out.println(">> Enter the image path (or 'Q'/'q' to quit):");
                String imagePath = scanner.nextLine();
                if (imagePath.isEmpty()) {
                    System.out.println("Invalid path.");
                    continue;
                }

                if (imagePath.equalsIgnoreCase("q")) {
                    return;
                }

                imagePath = imagePath.replaceAll("^\"|\"$", "");
                if (Files.notExists(Paths.get(imagePath))) {
                    System.out.println("The image path does not exist.");
                    continue;
                }

                int mode = 0;
                String templateName = null;
                while (true) {
                    System.out.println(">> Step 2: Choose a Mode Number:");
                    System.out.println("1. Read VIN from Barcode");
                    System.out.println("2. Read VIN from Text");

                    try {
                        mode = scanner.nextInt();
                        scanner.nextLine();
                        if (mode == 1 || mode == 2) {
                            templateName = mode == 1 ? "ReadVINBarcode" : "ReadVINText";
                            break;
                        }
                    } catch (Exception ignored) {
                        scanner.nextLine();
                    }
                }

                CapturedResult[] results = cvRouter.captureMultiPages(imagePath, templateName);
                if (results == null || results.length == 0) {
                    System.out.println("No results.");
                } else {
                    for (int index = 0; index < results.length; index++) {
                        CapturedResult result = results[index];
                        if (result.getErrorCode() == EnumErrorCode.EC_UNSUPPORTED_JSON_KEY_WARNING) {
                            System.out.println("Warning: " + result.getErrorCode() + ", " + result.getErrorString());
                        } else if (result.getErrorCode() != EnumErrorCode.EC_OK) {
                            System.out.println("Error: " + result.getErrorCode() + ", " + result.getErrorString());
                        }

                        ImageTag tag = result.getOriginalImageTag();
                        int pageNumber = tag instanceof FileImageTag ? ((FileImageTag)tag).getPageNumber() : index;

                        ParsedResult parsedResult = result.getParsedResult();
                        if (parsedResult == null || parsedResult.getItems().length == 0) {
                            System.out.println("Page-" + (pageNumber + 1) + " No parsed results.");
                        } else {
                            System.out.println("Page-" + (pageNumber + 1) + " Parsed.");
                            printResults(parsedResult);
                        }
                    }
                }
            }
        } finally {
            scanner.close();
        }
    }
}

File: samples/VIN/python/vin_scanner.py
import sys
from dynamsoft_capture_vision_bundle import *
import os

class VINResult:

    def __init__(self, item: ParsedResultItem):
        self.vin_string = None
        self.wmi = None
        self.region = None
        self.vds = None
        self.vis = None
        self.model_year = None
        self.plant_code = None
        self.code_type = item.get_code_type()
        if self.code_type != "VIN":
            return
        if item.get_field_value("vinString") != None:
            self.vin_string = item.get_field_value("vinString")
            if(item.get_field_validation_status("vinString")==EnumValidationStatus.VS_FAILED):
                self.vin_string += ", Validation Failed"

        if item.get_field_value("WMI") != None and item.get_field_validation_status("WMI") != EnumValidationStatus.VS_FAILED:
            self.wmi = item.get_field_value("WMI")

        if item.get_field_value("region") != None and item.get_field_validation_status("region") != EnumValidationStatus.VS_FAILED:
            self.region = item.get_field_value("region")

        if item.get_field_value("VDS") != None and item.get_field_validation_status("VDS") != EnumValidationStatus.VS_FAILED:
            self.vds = item.get_field_value("VDS")

        if item.get_field_value("VIS") != None and item.get_field_validation_status("VIS") != EnumValidationStatus.VS_FAILED:
            self.vis = item.get_field_value("VIS")

        if item.get_field_value("modelYear") != None and item.get_field_validation_status("modelYear") != EnumValidationStatus.VS_FAILED:
            self.model_year = item.get_field_value("modelYear")

        if item.get_field_value("plantCode") != None and item.get_field_validation_status("plantCode") != EnumValidationStatus.VS_FAILED:
            self.plant_code = item.get_field_value("plantCode")

    def to_string(self):
        return (f"VIN String: {self.vin_string}\n"
            f"Parsed Information:\n"
            f"\tWMI: {self.wmi or ''}\n"
            f"\tRegion: {self.region or ''}\n"
            f"\tVDS: {self.vds or ''}\n"
            f"\tVIS: {self.vis or ''}\n"
            f"\tModelYear: {self.model_year or ''}\n"
            f"\tPlantCode: {self.plant_code or ''}\n")

def print_results(result: ParsedResult) -> None:
    tag = result.get_original_image_tag()
    if isinstance(tag, FileImageTag):
        print("File:", tag.get_file_path())
    if result.get_error_code() != EnumErrorCode.EC_OK and result.get_error_code()!= EnumErrorCode.EC_UNSUPPORTED_JSON_KEY_WARNING:
        print("Error:", result.get_error_string())
    else:
        items = result.get_items()
        print("Parsed", len(items), "VIN codes.")
        for item in items:
            vin_result = VINResult(item)
            print(vin_result.to_string())

if __name__ == '__main__':

    print("**********************************************************")
    print("Welcome to Dynamsoft Capture Vision - VIN Sample")
    print("**********************************************************")

    # Initialize license.
    # You can request and extend a trial license from https://www.dynamsoft.com/customer/license/trialLicense?product=dcv&utm_source=samples&package=python
    # The string 'DLS2eyJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSJ9' here is a free public trial license. Note that network connection is required for this license to work.
    error_code, error_message = LicenseManager.init_license("DLS2eyJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSJ9")
    if error_code != EnumErrorCode.EC_OK and error_code != EnumErrorCode.EC_LICENSE_WARNING:
        print("License initialization failed: ErrorCode:", error_code, ", ErrorString:", error_message)
    else:
        cvr_instance = CaptureVisionRouter()
        while (True):
            image_path = input(">> Enter the image path (or 'Q'/'q' to quit):").strip(' \'"')
            if image_path == "":
                print("Invalid path.")
                continue

            if image_path.lower() == "q":
                sys.exit(0)

            if not os.path.exists(image_path):
                print("The image path does not exist.")
                continue
            while(True):
                print(">> Step 2: Choose a Mode Number:")
                print("1. Read VIN from Barcode")
                print("2. Read VIN from Text")
                try:
                    mode = int(input())
                    if mode == 1 or mode == 2:
                        template_name = "ReadVINBarcode" if mode == 1 else "ReadVINText"
                        break
                except ValueError:
                    continue

            result_array = cvr_instance.capture_multi_pages(image_path, template_name)
            results = result_array.get_results()
            if results is None or len(results) == 0:
                print("No results.")
            else:
                for i, result in enumerate(results):
                    page_number = i + 1
                    tag = result.get_original_image_tag()
                    if isinstance(tag, FileImageTag):
                        page_number = tag.get_page_number() + 1
                    if result.get_error_code() == EnumErrorCode.EC_UNSUPPORTED_JSON_KEY_WARNING:
                        print("Warning:", result.get_error_code(), result.get_error_string())
                    elif result.get_error_code() != EnumErrorCode.EC_OK:
                        print("Error:", result.get_error_code(), result.get_error_string())
                    parsed_result = result.get_parsed_result()

                    if parsed_result is None or len(parsed_result.get_items()) == 0:
                        print("Page-"+str(page_number), "No parsed results.")
                    else:
                        print_results(parsed_result)
                    print()
    input("Press Enter to quit...")
